File: tesseract/mirror/src/stores/workspace.ts
import { create } from 'zustand';
import { BACKEND_BASE } from '../lib/endpoints';

export type EventKind =
  | 'feedback_proposal'
  | 'feedback_sweep'
  | 'agent_approval'
  | 'skill_approval'
  | 'skill_refinement'
  | 'working_set_proposal'
  | 'soul_proposal'
  | 'change_proposal'
  | 'mission_reflection_proposal'
  | 'reflection_proposal'
  | 'nudge'
  | 'agent_post'
  | 'operator_post'
  | 'daily_brief'
  | 'yaml_change_proposal'
  | 'kb_merge_conflict'
  | 'clarification';

export type OperatorPostSource = 'button' | 'scratchpad' | 'voice' | 'hotkey' | 'telegram';

export type EventStatus = 'pending' | 'approved' | 'rejected' | 'resolved' | 'applied' | 'deleted';
export type Author = 'operator' | 'agent';

export interface WorkspaceComment {
  comment_id: string;
  event_id: string;
  ts: string;
  author: Author;
  body: string;
  reply_to: string | null;
  delivered_to_agent: boolean;
}

export interface WorkspaceEvent {
  event_id: string;
  ts: string;
  kind: EventKind;
  source: string;
  title: string;
  summary: string;
  payload: Record<string, unknown>;
  status: EventStatus;
  priority: number;
  decided_at: string | null;
  decided_reason: string | null;
  author_id: string;
  author_display: string;
  comments: WorkspaceComment[];
}

interface SeenMap {
  inbox?: string;
  stream?: string;
}

export type ThreadPendingState = 'queued' | 'thinking';

export interface ThreadPendingEntry {
  comment_id: string;
  state: ThreadPendingState;
}

interface WorkspaceState {
  events: WorkspaceEvent[];
  history: WorkspaceEvent[];
  seen: SeenMap;
  loading: boolean;
  loadingHistory: boolean;
  lastError: string | null;
  pendingThreads: Record<string, ThreadPendingEntry>;
  fetchInbox: () => Promise<void>;
  fetchHistory: () => Promise<void>;
  fetchSeen: () => Promise<void>;
  upsertEvent: (event: WorkspaceEvent) => void;
  appendComment: (comment: WorkspaceComment) => void;
  setThreadPending: (event_id: string, entry: ThreadPendingEntry | null) => void;
  refreshEvent: (event_id: string) => Promise<void>;
  /** Resolves `true` when the decision settled (applied, or the row was
   *  already gone) and `false` when it did not — a 5xx, a network drop, a
   *  parse failure. It reports rather than throws because a caller that only
   *  wants the side effect should not have to catch; the bulk caller needs
   *  the answer, and inferring it from a rejection is what a swallowing
   *  `catch` makes impossible. `lastError` still carries the message. */
  decide: (event_id: string, decision: 'approve' | 'reject' | 'resolve' | 'delete', reason?: string) => Promise<boolean>;
  comment: (event_id: string, body: string) => Promise<void>;
  markPanelSeen: (panel: 'inbox' | 'stream') => Promise<void>;
  unreadCount: () => number;
  attentionCount: () => number;
}

const ATTENTION_AGE_MS = 24 * 60 * 60 * 1000;

function pickNewer(local: string | undefined, remote: string | undefined): string | undefined {
  if (!local) return remote;
  if (!remote) return local;
  return local > remote ? local : remote;
}

class HttpError extends Error {
  status: number;
  // What the route said went wrong. One status covers several refusals here,
  // so the code alone cannot decide what to tell the operator.
  reason: string;
  constructor(path: string, status: number, reason = '') {
    super(`${path} ${status}`);
    this.status = status;
    this.reason = reason;
  }
}

async function failure(path: string, r: Response): Promise<HttpError> {
  const body = (await r.json().catch(() => ({}))) as { error?: unknown };
  return new HttpError(path, r.status, typeof body.error === 'string' ? body.error : '');
}

async function jget<T>(path: string): Promise<T> {
  const r = await fetch(`${BACKEND_BASE}${path}`);
  if (!r.ok) throw await failure(path, r);
  return r.json() as Promise<T>;
}

async function jpost<T>(path: string, body: unknown): Promise<T> {
  const r = await fetch(`${BACKEND_BASE}${path}`, {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(body),
  });
  if (!r.ok) throw await failure(path, r);
  return r.json() as Promise<T>;
}

function upsert(list: WorkspaceEvent[], event: WorkspaceEvent): WorkspaceEvent[] {
  const idx = list.findIndex((e) => e.event_id === event.event_id);
  if (idx < 0) return [event, ...list];
  const next = list.slice();
  next[idx] = event;
  return next;
}

/** Put one event on the side its status belongs to.
 *
 * `events` is the pending inbox and `history` is everything decided, so a row
 * whose status just changed has to move rather than be written where it was.
 * Deciding a row used to drop it from the inbox and the file watcher then
 * broadcast the same row back a quarter of a second later, which put it
 * straight back into the pending list: the operator saw an approved row
 * reappear, the badge kept counting it, and only a reload sorted it out.
 */
function place(
  state: { events: WorkspaceEvent[]; history: WorkspaceEvent[] },
  event: WorkspaceEvent,
): { events: WorkspaceEvent[]; history: WorkspaceEvent[] } {
  if (event.status === 'pending') {
    return {
      events: upsert(state.events, event),
      history: state.history.filter((e) => e.event_id !== event.event_id),
    };
  }
  return {
    events: state.events.filter((e) => e.event_id !== event.event_id),
    history: upsert(state.history, event),
  };
}

export const useWorkspaceStore = create<WorkspaceState>((set, get) => ({
  events: [],
  history: [],
  seen: {},
  loading: false,
  loadingHistory: false,
  lastError: null,
  pendingThreads: {},

  setThreadPending: (event_id, entry) =>
    set((s) => {
      const next = { ...s.pendingThreads };
      if (entry === null) {
        delete next[event_id];
      } else {
        next[event_id] = entry;
      }
      return { pendingThreads: next };
    }),

  fetchInbox: async () => {
    set({ loading: true, lastError: null });
    try {
      const data = await jget<{ events: WorkspaceEvent[] }>(
        '/api/workspace/inbox?status=pending',
      );
      set({ events: data.events, loading: false });
    } catch (err) {
      set({ loading: false, lastError: (err as Error).message });
    }
  },

  fetchHistory: async () => {
    set({ loadingHistory: true, lastError: null });
    try {
      const data = await jget<{ events: WorkspaceEvent[] }>(
        '/api/workspace/inbox?status=all',
      );
      const decided = data.events.filter((e) => e.status !== 'pending');
      set({ history: decided, loadingHistory: false });
    } catch (err) {
      set({ loadingHistory: false, lastError: (err as Error).message });
    }
  },

  fetchSeen: async () => {
    try {
      const data = await jget<SeenMap>('/api/workspace/seen');
      // Merge: keep whichever value is newer per panel. An optimistic
      // markPanelSeen call may have set a fresher local timestamp before
      // this fetch resolved — clobbering it would flicker the unread dots.
      set((s) => ({
        seen: {
          inbox: pickNewer(s.seen.inbox, data.inbox),
          stream: pickNewer(s.seen.stream, data.stream),
        },
      }));
    } catch {
      /* swallow — seen is best-effort */
    }
  },

  upsertEvent: (event) => set((s) => place(s, event)),

  appendComment: (comment) => {
    // Live-push from `workspace_comment_appended` envelope. Dedupe by
    // comment_id. If the parent event isn't in the in-memory list
    // (settled/filtered/never loaded), fall back to refreshEvent so
    // the broadcast isn't silently dropped — without this the operator
    // had to leave & re-enter the tab to see their own comment.
    const events = get().events.slice();
    const idx = events.findIndex((e) => e.event_id === comment.event_id);
    if (idx < 0) {
      void get().refreshEvent(comment.event_id);
      return;
    }
    const ev = events[idx];
    if (ev.comments.some((c) => c.comment_id === comment.comment_id)) return;
    events[idx] = { ...ev, comments: [...ev.comments, comment] };
    // Belt-and-suspenders: an assistant reply landing on the thread also clears
    // the `thinking…` indicator. Server fires `cleared` in `_run_turn`'s
    // finally, but a dropped envelope shouldn't leave the row spinning.
    if (comment.author === 'agent') {
      const pt = { ...get().pendingThreads };
      if (pt[comment.event_id]) {
        delete pt[comment.event_id];
        set({ events, pendingThreads: pt });
        return;
      }
    }
    set({ events });
  },

  refreshEvent: async (event_id) => {
    try {
      const ev = await jget<WorkspaceEvent>(`/api/workspace/event/${event_id}`);
      get().upsertEvent(ev);
    } catch {
      /* swallow — best-effort */
    }
  },

  decide: async (event_id, decision, reason) => {
    try {
      const updated = await jpost<WorkspaceEvent>(
        `/api/workspace/event/${event_id}/decision`,
        { decision, reason: reason ?? '' },
      );
      set((s) => place(s, updated));
      return true;
    } catch (err) {
      const status = err instanceof HttpError ? err.status : 0;
      if (status === 404) {
        // Stale row: backend purged it already. Drop locally so the
        // click registers as a removal; resync both panes silently.
        set((s) => ({
          events: s.events.filter((e) => e.event_id !== event_id),
          history: s.history.filter((e) => e.event_id !== event_id),
        }));
        void get().fetchInbox();
        void get().fetchHistory();
        // Settled from the operator's side: the row is gone either way.
        return true;
      }
      const reason = err instanceof HttpError ? err.reason : '';
      if (status === 409 && reason === 'concurrent_modification') {
        // The file moved after the proposal was written. The backend has
        // rewritten the card against the file as it stands now, so pull it
        // back in: what is on screen is what a second Approve would commit.
        // Narrowed on the reason, because the other refusals that share this
        // status are about the change itself, not about the file moving.
        void get().refreshEvent(event_id);
        set({
          lastError:
            'This file changed after the proposal was written. The card now shows the change against the file as it is now. Approve again to apply it.',
        });
        return false;
      }
      // Real failure (5xx, network, parse). Keep the row in place so
      // the operator can retry, surface the error so they can read it.
      set({ lastError: (err as Error).message });
      return false;
    }
  },

  comment: async (event_id, body) => {
    await jpost(`/api/workspace/event/${event_id}/comment`, { body });
    await get().refreshEvent(event_id);
  },

  markPanelSeen: async (panel) => {
    const last_seen_at = new Date().toISOString();
    set({ seen: { ...get().seen, [panel]: last_seen_at } });
    try {
      await jpost('/api/workspace/seen', { panel, last_seen_at });
    } catch {
      /* swallow */
    }
  },

  unreadCount: () => {
    // Inbox is fetched with `?status=pending`, so `events` is already the
    // "needs operator decision" set. The previous timestamp-vs-lastSeen
    // semantic dropped the badge to 0 the moment the operator opened the
    // tab — even though the items were still awaiting action. The HUD
    // badge means "you have stuff to do here," not "things arrived since
    // your last visit," so we count the pending set directly.
    return get().events.length;
  },

  attentionCount: () => {
    // m2 (Codex 2026-05-06): operator_post now has a real `resolve`
    // verb, so the previous "skip operator_post in attention aging"
    // workaround is gone. An unresolved operator thread sitting >24h
    // is a real stuck item the operator should see.
    const now = Date.now();
    return get().events.filter((e) => {
      const t = Date.parse(e.ts);
      return Number.isFinite(t) && now - t > ATTENTION_AGE_MS;
    }).length;
  },
}));

File: tesseract/mirror/server/routes/workspace.py
"""Workspace REST routes — Inbox + comment threads.

Endpoints:

- ``GET  /api/workspace/inbox``                — pending events (+ filter)
- ``GET  /api/workspace/event/{event_id}``     — single event + thread
- ``POST /api/workspace/event/{event_id}/decision``  — approve / reject
- ``POST /api/workspace/event/{event_id}/comment``   — append operator comment
- ``GET  /api/workspace/seen``                 — last-seen markers
- ``POST /api/workspace/seen``                 — update last-seen marker

The document-editor endpoints (`/api/workspace/docs`, `/api/workspace/doc`)
live in `workspace_docs.py`, mounted by `app.py` alongside these.

A future ``/api/workspace/stream`` would read the same events.jsonl
with a ``kind=stream`` filter.

Approve dispatcher
==================

This is the decision dispatcher: `apply_decision` and the HTTP shape around
it. The commit handlers it calls are split across sibling modules by domain
— `workspace_skill_commits.py` for `agent_approval` and the three
`skill_*` kinds, `workspace_proposal_commits.py` for the file/memory-edit
kinds, `workspace_dial_commits.py` for the two dial cards (`tuning_proposal`,
`working_set_proposal`). Every `_commit_*` name is imported here because
`apply_decision` calls it directly, which is also what keeps
`tesseract.mirror.server.routes.workspace._commit_skill_refinement` (etc.)
importable for existing tests without a separate re-export step.

Approve on a `change_proposal` event triggers a kind-aware commit via
`tesseract.kernel.workspace_changes.apply_change`. Concurrent
modification (file changed since the proposal was queued) returns 409
with the fresh diff so the operator can re-review.

Approve on a `soul_proposal` event (feedback consolidator's distilled
identity bullet) commits the bullet directly to SOUL.md's `Growth`
section — the consolidator card already shows the bullet text, so the
single Approve click is the operator's sole gate. The intermediate
``change_proposal`` step is skipped to avoid double-gating the same
decision.

Approve on a `feedback_proposal` or `feedback_sweep` event invokes the
`memory_promote` or `memory_save` tool with the payload's action — operator
decision is the gate, the memory store mutation runs synchronously, status
flips to ``applied``.
"""

from __future__ import annotations

import asyncio
import logging
from typing import Any

from aiohttp import web

from tesseract.mirror.server.routes.workspace_dial_commits import (
    _commit_tuning_proposal,
    _commit_working_set_proposal,
)
from tesseract.mirror.server.routes.workspace_proposal_commits import (
    _commit_change_proposal,
    _commit_feedback_proposal,
    _commit_soul_proposal,
    _commit_vault_raw_ingest_batch,
    _commit_yaml_change_proposal,
)
from tesseract.mirror.server.routes.workspace_skill_commits import (
    _apply_skill_refinement,  # re-exported: called directly by existing tests
    _base_moved,  # re-exported: called directly by existing tests
    _commit_agent_approval,
    _commit_skill_approval,
    _commit_skill_refinement,
    _commit_skill_retirement,
    _subject_of,
)
from tesseract.mirror.server.routes.workspace_shared import _store
from tesseract.lib import last_seen
from tesseract.paths import ROOT  # re-exported: `_commit_yaml_change_proposal`
# moved to `workspace_proposal_commits.py` and reads its own `ROOT` there;
# several existing tests still monkeypatch `ws_routes.ROOT` defensively (they
# do not exercise `yaml_change_proposal`, so the patch is inert either way).
from tesseract.permissions.approval_log import record_ask
from tesseract.workspace_events import (
    WorkspaceComment,
    WorkspaceEvent,
)
from tesseract.workspace_events.events import RESOLVABLE_KINDS

log = logging.getLogger(__name__)


# `resolve` is the soft-close verb for informational events — threads
# that record something that already happened (the assistant post, dream-cycle
# nudge, operator-initiated thread) or where the system has nothing
# left to gate (session reflection: writes already committed during
# the reflect turn). Derived from `workspace_events.events.ANSWERABLE_WITH`
# rather than kept as its own list — see that mapping's docstring for
# why three separate copies of this is how a kind ships stuck.
#
# Gated kinds — `change_proposal`, `feedback_proposal`, `feedback_sweep`,
# `agent_approval`, `soul_proposal`, `mission_reflection_proposal` —
# MUST go through explicit approve/reject so the operator's decision
# is recorded; allowing `resolve` on those would erase the gate.
_RESOLVABLE_KINDS = RESOLVABLE_KINDS

# Per-event decision locks. `post_decision` reads the event, runs commit
# side-effects, then flips its status — a check-then-act span. Two concurrent
# Approves on the same event could both pass the pending check and both run the
# commit (the second then 409s on hash mismatch — invariant intact but
# confusing). A per-event_id asyncio.Lock serializes that span; different
# events never contend. The dict is not pruned — bounded by distinct decided
# event_ids (small for an operator inbox), and pruning a lock with a pending
# waiter would let a fresh request mint a second lock and re-open the race.
_decision_locks: dict[str, asyncio.Lock] = {}


def _decision_lock(event_id: str) -> asyncio.Lock:
    lock = _decision_locks.get(event_id)
    if lock is None:
        lock = asyncio.Lock()
        _decision_locks[event_id] = lock
    return lock


def _event_dict(ev: WorkspaceEvent, comments: list[WorkspaceComment]) -> dict[str, Any]:
    return {
        **ev.to_dict(),
        "comments": [c.to_dict() for c in comments],
    }


async def list_inbox(request: web.Request) -> web.Response:
    app = request.app
    store = _store(app)
    status_filter = request.query.get("status", "pending")
    status: Any = status_filter if status_filter != "all" else None
    events = store.list_events(status=status, limit=200)
    payload = []
    for ev in events:
        comments = store.list_comments(ev.event_id)
        payload.append(_event_dict(ev, comments))
    return web.json_response({"events": payload, "count": len(payload)})


async def get_event(request: web.Request) -> web.Response:
    app = request.app
    store = _store(app)
    event_id = request.match_info["event_id"]
    ev = store.get_event(event_id)
    if ev is None:
        return web.json_response({"error": "not_found"}, status=404)
    comments = store.list_comments(event_id)
    return web.json_response(_event_dict(ev, comments))


class DecisionError(Exception):
    """A decision that could not be made, with the reason a caller can render.

    Carries the same payload and status the route used to return inline, so the
    HTTP handler is a translation and every other caller gets the words rather
    than a status code it has no use for.
    """

    def __init__(self, payload: dict[str, Any], status: int) -> None:
        self.payload = payload
        self.status = status
        super().__init__(str(payload.get("detail") or payload.get("error") or "refused"))


async def apply_decision(
    app: web.Application,
    event_id: str,
    decision: str,
    *,
    reason: str | None = None,
    per_file: dict[str, str] | None = None,
    actor: str = "operator",
) -> tuple[WorkspaceEvent, list[WorkspaceComment]]:
    """Approve, reject, resolve or delete one workspace event.

    **The one place a decision happens, and it is no longer behind HTTP.**
    Everything the operator is asked in the cockpit lands in this inbox, and
    until this was liftable out of `post_decision` the only way to answer any
    of it was a `POST` from a browser on the same machine. Ruling 22 says every
    approval has to be answerable from a phone; measured, two of fourteen were,
    and the twelve that were not are all this one call. So it moved, unchanged,
    and the route below is now the HTTP shape around it rather than the thing
    itself.

    Raises `DecisionError` with the payload and status the route used to return.
    """
    store = _store(app)
    if decision not in {"approve", "reject", "resolve", "delete"}:
        raise DecisionError(
            {"error": "decision must be 'approve', 'reject', 'resolve', or 'delete'"},
            400,
        )

    ev = store.get_event(event_id)
    if ev is None:
        raise DecisionError({"error": "not_found"}, 404)
    # `resolve` is restricted to informational kinds. A pending
    # `change_proposal` / `mission_reflection_proposal` / feedback_* /
    # agent_approval MUST go through approve/reject so the decision is
    # recorded; a blanket `resolve` on those would silently bypass the gate.
    if decision == "resolve" and ev.kind not in _RESOLVABLE_KINDS:
        raise DecisionError(
            {
                "error": "resolve_not_permitted_for_kind",
                "detail": (
                    f"'resolve' is only valid for {sorted(_RESOLVABLE_KINDS)}; "
                    f"event kind '{ev.kind}' requires explicit approve or reject."
                ),
                "kind": ev.kind,
            },
            400,
        )

    async with _decision_lock(event_id):
        # Re-read under the lock. A concurrent decision may have settled it
        # while we waited; acting on the stale copy is the race this guards.
        ev = store.get_event(event_id)
        if ev is None:
            raise DecisionError({"error": "not_found"}, 404)
        # `delete` is the universal escape hatch — works on any status so the
        # operator can soft-delete from history too. Other verbs stay
        # pending-only and idempotently return the settled event.
        if decision != "delete" and ev.status not in {"pending"}:
            return ev, store.list_comments(event_id)

        applied_version = ""

        if decision == "approve" and ev.kind == "change_proposal":
            err = await _commit_change_proposal(app, ev)
            if err is not None:
                raise DecisionError(*err)

        if decision == "approve" and ev.kind == "yaml_change_proposal":
            _result, err = await _commit_yaml_change_proposal(app, ev)
            if err is not None:
                raise DecisionError(*err)

        if decision == "approve" and ev.kind == "soul_proposal":
            _result, err = await _commit_soul_proposal(app, ev)
            if err is not None:
                raise DecisionError(*err)

        if decision == "approve" and ev.kind == "working_set_proposal":
            _result, err = await _commit_working_set_proposal(app, ev)
            if err is not None:
                raise DecisionError(*err)

        if decision == "approve" and ev.kind == "tuning_proposal":
            _result, err = await _commit_tuning_proposal(app, ev)
            if err is not None:
                raise DecisionError(*err)

        # feedback_sweep reaches the same handler as feedback_proposal: its
        # payload names `action: memory_save` rather than `merge_into`/`archive`, and `_commit_feedback_proposal`
        # dispatches on that field.
        if decision == "approve" and ev.kind in {"feedback_proposal", "feedback_sweep"}:
            _result, err = await _commit_feedback_proposal(app, ev)
            if err is not None:
                raise DecisionError(*err)

        if decision in {"approve", "reject"} and ev.kind == "agent_approval":
            _agent_meta, err = await _commit_agent_approval(app, ev, decision, reason)
            if err is not None:
                raise DecisionError(*err)

        if decision in {"approve", "reject"} and ev.kind == "skill_approval":
            _skill_meta, err = await _commit_skill_approval(app, ev, decision, reason)
            if err is not None:
                raise DecisionError(*err)

        if decision in {"approve", "reject"} and ev.kind == "skill_refinement":
            refine_meta, err = await _commit_skill_refinement(app, ev, decision, reason)
            if err is not None:
                raise DecisionError(*err)
            applied_version = str((refine_meta or {}).get("applied_version") or "")

        if decision in {"approve", "reject"} and ev.kind == "skill_retirement":
            _retirement_meta, err = await _commit_skill_retirement(app, ev, decision, reason)
            if err is not None:
                raise DecisionError(*err)

        if decision in {"approve", "reject"} and ev.kind == "vault_raw_ingest_batch":
            _result, err = await _commit_vault_raw_ingest_batch(
                app, ev, deny_all=(decision == "reject"), per_file=per_file
            )
            if err is not None:
                raise DecisionError(*err)

        # `resolve` flips status without firing approval side-effects — the
        # verb operator_post threads need to leave the inbox once the
        # conversation has played out. `delete` is the soft-delete verb.
        # `applied` distinguishes "the operator approved AND the side-effect
        # ran" from the generic approval state.
        if decision == "approve" and ev.kind in {
            "yaml_change_proposal",
            "vault_raw_ingest_batch",
            "soul_proposal",
            "feedback_proposal",
            "feedback_sweep",
            "agent_approval",
            "skill_approval",
            "skill_refinement",
            "skill_retirement",
            "working_set_proposal",
            "tuning_proposal",
        }:
            new_status = "applied"
        else:
            new_status = (
                "approved" if decision == "approve"
                else "rejected" if decision == "reject"
                else "resolved" if decision == "resolve"
                else "deleted"
            )
        updated = store.update_event_status(event_id, new_status, reason=reason)
        if updated is None:
            raise DecisionError({"error": "not_found"}, 404)

    try:
        await record_ask(
            session_id="workspace",
            call_id=event_id,
            tool_name="workspace_decision",
            input_summary={
                "kind": ev.kind,
                "decision": decision,
                "reason": reason or "",
                # What the decision was ABOUT. Without these the ledger holds
                # one row per settled card carrying the kind and the verdict
                # and nothing that names the thing, so "what changed this
                # playbook, and did the change help" cannot be asked of it.
                # `_subject_of` returns {} for a kind that has no subject, so
                # the row is unchanged for every other card.
                **_subject_of(ev, applied_version),
            },
            posture_source="workspace_decision",
            result=(
                "allow_once" if decision == "approve"
                else "resolved" if decision == "resolve"
                else "deleted" if decision == "delete"
                else "deny"
            ),
            # Who actually decided. It was hardcoded to the operator, which
            # was true while a person was the only thing that could reach
            # here, and is a lie the moment the mode settles a card itself.
            # The approval ledger is the record of who allowed what.
            actor=actor,
        )
    except Exception:
        log.exception("workspace: approval ledger record failed")

    return updated, store.list_comments(event_id)


async def post_decision(request: web.Request) -> web.Response:
    """The HTTP shape around `apply_decision`, and nothing else."""
    event_id = request.match_info["event_id"]
    try:
        body = await request.json()
    except Exception:
        return web.json_response({"error": "invalid_json"}, status=400)

    raw = body.get("decisions") if isinstance(body, dict) else None
    try:
        updated, comments = await apply_decision(
            request.app,
            event_id,
            (body.get("decision") or "").strip().lower(),
            reason=(body.get("reason") or "").strip() or None,
            per_file=raw if isinstance(raw, dict) else None,
        )
    except DecisionError as exc:
        return web.json_response(exc.payload, status=exc.status)
    # A card answered is the operator being here, and for a day spent reading
    # the inbox rather than talking it is the only trace of it. On the route
    # rather than inside `apply_decision`, because the other caller is a tool
    # the assistant runs, and under `free` that tool answers unattended.
    last_seen.record()
    return web.json_response(_event_dict(updated, comments))


async def post_comment(request: web.Request) -> web.Response:
    app = request.app
    store = _store(app)
    event_id = request.match_info["event_id"]
    try:
        body = await request.json()
    except Exception:
        return web.json_response({"error": "invalid_json"}, status=400)

    text = (body.get("body") or "").strip()
    if not text:
        return web.json_response({"error": "body required"}, status=400)
    if store.get_event(event_id) is None:
        return web.json_response({"error": "not_found"}, status=404)

    comment = WorkspaceComment.new(
        event_id=event_id,
        author="operator",
        body=text,
        reply_to=body.get("reply_to") or None,
    )
    store.append_comment(comment)
    # Talking in a thread is being here as much as answering a card is. A day
    # spent reading the inbox and replying, without settling anything, would
    # otherwise leave the operator marked absent for all of it.
    last_seen.record()

    # Live-push the operator comment to all attached Mirror sessions so the
    # CommentThread renders without a manual refresh. Best-effort — the
    # comment is durable on disk regardless.
    try:
        from tesseract.workspace_events.broadcast import broadcast_comment_appended
        await broadcast_comment_appended(app, comment)
    except Exception:
        log.exception("workspace: broadcast_comment_appended failed")

    # Dispatch a controller session to write the assistant reply directly into
    # the workspace thread (durable — controller calls workspace_reply tool
    # before this returns). Backend reads + broadcasts the controller-written
    # comment; never writes the reply itself. spawn_if_missing=False so no
    # daemon is cold-forked from a web request.
    try:
        from tesseract.mirror.server.ws import _spawn_tracked
        from tesseract.orchestrator.autonomy.workspace_reply_dispatch import (
            dispatch_workspace_reply,
            load_workspace_reply_config,
        )
        ev = store.get_event(event_id)
        cfg = load_workspace_reply_config()
        if cfg.enabled and ev is not None:
            _spawn_tracked(
                app,
                dispatch_workspace_reply(
                    app,
                    event_id=event_id,
                    comment_id=comment.comment_id,
                    event=ev,
                    kind="comment",
                    comment_text=text,
                    config=cfg,
                ),
                name=f"workspace-reply:{event_id}",
            )
    except Exception:
        log.exception("workspace: failed to spawn workspace reply")

    return web.json_response(comment.to_dict(), status=201)


_OPERATOR_POST_SOURCES = {"button", "scratchpad", "voice", "hotkey", "telegram"}
_OPERATOR_POST_TITLE_MAX = 200
_OPERATOR_POST_BODY_MAX = 4000


async def post_operator_post(request: web.Request) -> web.Response:
    """Workstream D — operator-initiated workspace thread.

    Body: ``{title, body, source}``. Optional query ``?await_reply=false``
    suppresses the synthetic turn (default fires it so the operator gets
    an assistant reply within seconds without manually leaving a comment).
    """
    app = request.app
    store = _store(app)
    try:
        body = await request.json()
    except Exception:
        return web.json_response({"error": "invalid_json"}, status=400)

    title = (body.get("title") or "").strip()
    text = (body.get("body") or "").strip()
    source = (body.get("source") or "").strip()
    if not text:
        return web.json_response({"error": "body required"}, status=400)
    if source not in _OPERATOR_POST_SOURCES:
        return web.json_response(
            {"error": f"source must be one of {sorted(_OPERATOR_POST_SOURCES)}"},
            status=400,
        )
    if len(title) > _OPERATOR_POST_TITLE_MAX:
        title = title[:_OPERATOR_POST_TITLE_MAX]
    if len(text) > _OPERATOR_POST_BODY_MAX:
        text = text[:_OPERATOR_POST_BODY_MAX]
    if not title:
        # Derive a one-line title from the body so the inbox row is
        # legible. Strip newlines so the row doesn't grow vertically.
        first_line = text.splitlines()[0] if text else ""
        title = (first_line[:80] or "Operator note").strip()

    event = WorkspaceEvent.new(
        kind="operator_post",
        source="operator",
        title=title,
        summary=text[:400],
        payload={"body": text, "source": source},
    )
    store.append_event(event)
    last_seen.record()

    try:
        from tesseract.workspace_events.broadcast import broadcast_workspace_event
        await broadcast_workspace_event(app, event)
    except Exception:
        log.exception("workspace: broadcast_workspace_event (operator_post) failed")

    await_reply = request.query.get("await_reply", "true").lower() != "false"
    if await_reply:
        try:
            from tesseract.mirror.server.ws import _spawn_tracked
            from tesseract.orchestrator.autonomy.workspace_reply_dispatch import (
                dispatch_workspace_reply,
                load_workspace_reply_config,
            )
            cfg = load_workspace_reply_config()
            if cfg.enabled:
                _spawn_tracked(
                    app,
                    dispatch_workspace_reply(
                        app,
                        event_id=event.event_id,
                        comment_id=event.event_id,
                        event=event,
                        kind="post",
                        comment_text=text,
                        config=cfg,
                    ),
                    name=f"workspace-reply:{event.event_id}",
                )
        except Exception:
            log.exception("workspace: failed to spawn workspace post reply")

    return web.json_response(event.to_dict(), status=201)


async def get_seen(request: web.Request) -> web.Response:
    app = request.app
    store = _store(app)
    return web.json_response(store.get_seen())


async def post_seen(request: web.Request) -> web.Response:
    app = request.app
    store = _store(app)
    try:
        body = await request.json()
    except Exception:
        return web.json_response({"error": "invalid_json"}, status=400)
    panel = (body.get("panel") or "").strip()
    if panel not in {"inbox", "stream"}:
        return web.json_response(
            {"error": "panel must be 'inbox' or 'stream'"}, status=400,
        )
    ts = (body.get("last_seen_at") or "").strip()
    if not ts:
        return web.json_response({"error": "last_seen_at required"}, status=400)
    store.set_seen(panel, ts)
    return web.json_response({"ok": True, "panel": panel, "last_seen_at": ts})
